nixos/modules/security/wrappers/wrapper.c

```diff
@@ -125,113 +126,120 @@ static int make_caps_ambient(const char *self_path) {
+        return NULL;
+    }
+
+    if (elf_size == file_size) {
+        fprintf(stderr, LOG_PREFIX "no extra bytes found for target program name found in security wrapper, did you forget to concat it?\n");
+        return NULL;
+    }
+
+    char *prog_name = calloc(file_size - elf_size, 1);
+    if (!prog_name) {
+        perror(LOG_PREFIX "cannot allocate memory for prog_name");
+        return NULL;
+    }
+
+    if (pread_all(exe_fd, prog_name, file_size - elf_size, elf_size) == -1) {
+        free(prog_name);
+        return NULL;
+    };
+
+    return prog_name;
 }
 
 int main(int argc, char **argv) {
-    ASSERT(argc >= 1);
-    char *self_path = NULL;
-    int self_path_size = readlink_malloc("/proc/self/exe", &self_path);
-    if (self_path_size < 0) {
-        fprintf(stderr, "cannot readlink /proc/self/exe: %s", strerror(-self_path_size));
-    }
-
-    // Make sure that we are being executed from the right location,
-    // i.e., `safe_wrapper_dir'.  This is to prevent someone from creating
-    // hard link `X' from some other location, along with a false
-    // `X.real' file, to allow arbitrary programs from being executed
-    // with elevated capabilities.
-    int len = strlen(wrapper_dir);
-    if (len > 0 && '/' == wrapper_dir[len - 1])
-      --len;
-    ASSERT(!strncmp(self_path, wrapper_dir, len));
-    ASSERT('/' == wrapper_dir[0]);
-    ASSERT('/' == self_path[len]);
-
-    // Make *really* *really* sure that we were executed as
-    // `self_path', and not, say, as some other setuid program. That
-    // is, our effective uid/gid should match the uid/gid of
-    // `self_path'.
-    struct stat st;
-    ASSERT(lstat(self_path, &st) != -1);
-
-    ASSERT(!(st.st_mode & S_ISUID) || (st.st_uid == geteuid()));
-    ASSERT(!(st.st_mode & S_ISGID) || (st.st_gid == getegid()));
-
-    // And, of course, we shouldn't be writable.
-    ASSERT(!(st.st_mode & (S_IWGRP | S_IWOTH)));
-
-    // Read the path of the real (wrapped) program from <self>.real.
-    char real_fn[PATH_MAX + 10];
-    int real_fn_size = snprintf(real_fn, sizeof(real_fn), "%s.real", self_path);
-    ASSERT(real_fn_size < sizeof(real_fn));
-
-    int fd_self = open(real_fn, O_RDONLY);
-    ASSERT(fd_self != -1);
-
-    char source_prog[PATH_MAX];
-    len = read(fd_self, source_prog, PATH_MAX);
-    ASSERT(len != -1);
-    ASSERT(len < sizeof(source_prog));
-    ASSERT(len > 0);
-    source_prog[len] = 0;
-
-    close(fd_self);
+    int exe_fd = open("/proc/self/exe", O_CLOEXEC);
+    if (exe_fd < 0) {
+        perror(LOG_PREFIX "cannot open /proc/self/exe");
+        return 1;
+    }
+
+    char *prog_name = get_progname(exe_fd);
+    if (!prog_name) {
+        return 1;
+    }
 
     // Read the capabilities set on the wrapper and raise them in to
     // the ambient set so the program we're wrapping receives the
     // capabilities too!
-    if (make_caps_ambient(self_path) != 0) {
-        free(self_path);
+    if (make_caps_ambient(exe_fd) != 0) {
```

Comment:
Hm, if we allow just-execute under no-new-priv, should we just log a warning here and go on anyway?

Reply:
According to capabilities(7) no, because we might have raised some, but not all the capabilities we wanted raised.

Reply:
Well, wouldn't it be more consistent with the other case to go on and see if the capabilities we can get are enough anyway for the task the user wants, though?

Reply:
Maybe. However, the documentation of capabilitysets is very confusing and claims that it's somehow important not to run things with subsets of capabilities they were intended to run with. I would like to not change the behaviour here until we understand why capabilities(7) makes such statements.

Reply:
Hmmm. Wording looks generic in the «race-condition-like-assumption-breaking-style»…

Reply:
Take a look at "Safety checking for capability-dumb binaries" section of [capabilities(7)](https://man7.org/linux/man-pages/man7/capabilities.7.html). I don't know if things that section says are not very misleading (that manpage is either self-contradictory or very hard to understand: https://bugzilla.kernel.org/show_bug.cgi?id=216667), but if not it implies that setting some-but-not-all capabilities could be a security issue for some targets.

Reply:
I have looked at it before the previous replies, that's why I said «wording looks like».

What is even more annoying is the following lack of clarity: so kernel will not let you create some specific issue with `execve`; but will it ever let you create this issue (or will setcap just -EPERM without changing the capabilities, quick look at the kernel source does look like that…)

Reply:
In the filecap case, it's unclear which "you" did you mean. There is whoever set the filecap, and also whoever executes the filecap binary.

Whoever has the ability to set filecaps on a file can already exercise them however they wish, so they gain nothing from being able to confuse a binary that doesn't have any other filecaps. They are in a similar position to someone who runs with those caps and then can mess with them using setcap -- they already have them and so confusing anyone _who doesn't have additional permissions_ is not an escalation.

On the other hand it would be a potential escalation if someone who can run a filecap binary could cause it to be confused in that way, because when they run that binary it has some authority that they don't have.

Reply:
You whoever plays with a system to make a point to be honest (i.e. you do have root to setup a silly situation with containers and setcap and seccomp and whatever else should not be mixed, and then as a non-root user you do something you are not supposed to be able to do!)

It just looks like the same invariants are enforced on execve and on changing the capability sets, but I did not look carefully so I definitely missed something (the question is whether it makes any difference). 

If we consider things like SGID + setcap and having one but not the other applied confuses the wrapped executable, we should also fail on setuid/setgid failure, no?

Reply:
> If we consider things like SGID + setcap and having one but not the other applied confuses the wrapped executable, we should also fail on setuid/setgid failure, no?

No. There is no way to partially setgid.

Reply:
There is no way to «partially setgid», but setcap-without-setgid might still be a partial application of intended security context.

Reply:
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/security/wrappers/default.nix#L218

Reply:
Ah this exclusivity is still there! Good, I guess

> If so, then even the interface is different (according to docs): there is supposedly no effective capset on files, but only a single effective bit that applies to all caps.

Hm, this part is not that interesting (different interfaces ending up with same conditions are fine), but there is indeed a capability-by-capability operation involved in the wrapper, so partial change is possible, you are right.